blog/_posts/2021-03-15-scaling-inlets.md

```diff
@@ -0,0 +1,349 @@
+>
+> ``` bash
+> #!/bin/bash
+> for i in {1..50}
+> do
+>   n=`printf %03d $i`
+>   ./create-exit-server.sh "app-$n" 8080 &
+> done
+> 
+> wait
+> ```
+
+## Connecting the client
+
+Now that the server part of the tunnel is ready, connect a client to bring an application into the remote cluster.
+
+First, grab the token of the target tunnel:
+
+``` bash
+export TOKEN=$(kubectl get secrets inlets-keycloak-token --template={{.data.token}} | base64 -d)
+```
+
+Now connect your inlets client to the inlets server:
+
+- `--url`: the secure websocket url
+- `--auto-tls=false`: don't get the certificate from the control plane, a valid Let's Encrypt certificate is in use
+- `--upstream` and `--port`: the target host and port
+- `--license-file`: a valid inlets PRO license
+
+``` bash
+inlets-pro tcp client \
+  --url wss://keycloak-tunnel.inlets.example.com \
+  --auto-tls=false \
+  --upstream=localhost \
+  --port=8080 \
+  --license-file=$HOME/.inlets/license
+```
+
+## Exposing the data plane
+
+By default, the data plane is only reachable from within the Kubernetes cluster. This could be enough for many use case like the Prometheus monitoring setup, or when you're building a SaaS platform and you want your client to make a service on-premise available in your platform.
+
+But what if you want to expose this application to the outside world as well? What options do we have?
+
+You could change the service type of the data plane to a LoadBalancer, but this will create extra cloud resources for each tunnel you want to expose. Not only will it create extra costs, but there are also limits in place when creating hundreds or thousands of such load balancer.
+
+In the current set up, nginx ingress controller and cert-manager is already installed, so if the application is HTTP based, like Keycloak, Wordpress or Jenkins, an extra ingress could expose the service to the public. E.g:
+
+``` yaml
+apiVersion: networking.k8s.io/v1
+kind: Ingress
+metadata:
+  name: keycloak-data-plane
+spec:
+  rules:
+  - host: keycloak.inlets.example.com
+    http:
+      paths:
+      - backend:
+          service:
+            name: keycloak-inlets-pro-data-plane
+            port:
+              number: 8080
+        path: /
+  tls:
+  - hosts:
+    - keycloak.inlets.example.com
+    secretName: inlets-tls
+```
+
+## Wrapping up
+
+In this post we started by creating a single inlets PRO exit server in a Kubernetes cluster with the Helm chart, but instead of using the default ingress and cert-manager configuration we issued a wildcard certificate to support multiple tunnels.
+We created a first tunnel with a little utility script, making it easy to configure some settings like the data plane port and the authentication token. Next, we used the same script to add more and more tunnels.
+
+It should give you an idea how easy it is to scale up to hundreds of tunnels by using the Helm chart.
```

Comment:
Can you refer to the first blog post where I introduce inlets-cloud and use its diagram here too, to say that it's available to automate and manage everything you showed in the post?

Reply:
https://inlets.dev/blog/2020/10/08/advanced-cloud-patterns.html